include/jemalloc/internal/arena_inlines_b.h

```diff
@@ -148,23 +148,23 @@ arena_decay_tick(tsdn_t *tsdn, arena_t *arena) {
 
 JEMALLOC_ALWAYS_INLINE void *
 arena_malloc(tsdn_t *tsdn, arena_t *arena, size_t size, szind_t ind, bool zero,
-    tcache_t *tcache, bool slow_path) {
+    bool slab, tcache_t *tcache, bool slow_path) {
 	assert(!tsdn_null(tsdn) || tcache == NULL);
 
 	if (likely(tcache != NULL)) {
-		if (likely(size <= SC_SMALL_MAXCLASS)) {
+		if (likely(size <= SC_SMALL_MAXCLASS && slab)) {
 			return tcache_alloc_small(tsdn_tsd(tsdn), arena,
 			    tcache, size, ind, zero, slow_path);
 		}
-		if (likely(size <= tcache_maxclass)) {
+		if (likely(size <= tcache_maxclass && !slab)) {
```

Comment:
This looks wrong? When sampled, we may ask for size == 4K && !slab, which will be served using tcache_alloc_large incorrectly?

Reply:
Good catch, thanks.